content/blog/sivayanama-testing-opsramp-apis-with-pytest-fixtures.md

```diff
@@ -0,0 +1,232 @@
+
+```python
+@pytest.fixture(autouse=True)
+def get_auth_header(api_endpoint, client_id, client_secret, grant_type="client_credentials"):
+```
+
+
+
+## Provisioning of third-party plugin Fixtures 
+
+Not necessarily have to define all fixtures on your own code. Fixtures can be provided by third-party plugins as well and you are free to use them. You just need to install the required plugins. We will see an example with pytest-datadir plugin.
+
+
+
+Install this plugin as shown below 
+
+```shell
+pip install pytest-datadir
+```
+
+
+
+```python
+@pytest.fixture(autouse=True)
+
+def get_bearer_token(shared_datadir,request):
+
+json_path = (shared_datadir / "auth_header_input.json")
+```
+
+![](https://lh4.googleusercontent.com/dacTgDdw17BzeyCitShA73WSip9LVtenQoNN-uraaN5tKEU5cA_xP3cEmNPWmTzU3A1HegdoOVvwPbyYqQuoLeEk4W766nIvpBdoTzUdIiT2dXiOQG0_h7atQWS7-T9qvRrieuhlEV84VS15ir11Ocw)
+
+
+
+get\_bearer\_token fixture is using [shared_datadir](https://pypi.org/project/pytest-datadir/) fixture from third party plugin. Shared_datadir fixture provision data folder from the current folder as pathlib.Path object. 
+
+
+
+## Fixture finalization
+
+Usually, you do teardown or cleanup activities for each test for many good reasons such as the following: 
+
+- The executed tests will not impact other tests
+
+- Do not want to have the testing environment piled up with tons of test data
+
+- Every test should execute in a clean state
+
+
+
+Fixtures offer a very powerful teardown, cleanup capability known as finalization. 
+
+
+```python
+@pytest.fixture()
+
+def create_client( shared_datadir, get_bearer_token,request):
+
+client = Client()
+
+client.logger.debug("IN\*\**")
+
+
+
+payload_file = (shared_datadir / "create_client1.json")
+
+input_payload= client.http_helper.to_json(payload_file)
+
+
+
+client_response = client.create_client(input_payload,get_bearer_token)
+
+
+
+def terminate_client():
+
+client.logger.debug("IN\*\**")
+
+util=Utils()
+
+
+
+clientId=Client.clientId
+
+
+
+if len(str(clientId)) > 0:
+
+oauth_token=get_bearer_token
+
+client.terminate_client(clientId, oauth_token)
+
+client.logger.debug("OUT\*\**")
+
+
+
+request.addfinalizer(terminate_client)
+
+
+
+
+
+client.print_log(client_response)
+
+client.logger.debug("OUT\*\**")
+
+return client_response
+```
+
+You should add the finalizer function to the request’s context object of the test as shown below:
+
+*request.addfinalizer(terminate_client)*
+
+
+
+In the above example, you do all the test setup on the client and execute the test. Once the test is complete finalizer function runs and cleans up the whole thing. 
+
+
+
+## Conclusion
+
+In this article, you have seen what are PyTest Fixtures, and how to define, invoke, and Fixtures. Also, you have seen how to leverage testing of OpsRamp APIs using these fixtures, how to do teardown activities using finalizer functions. 

```

Comment:
Consider this rephrasing:
"In this article, I have explained what PyTest fixtures are and how to define, invoke and implement fixtures. I also showed you how to leverage the testing of OpsRamp APIs using these fixtures and how to do teardown activities using the finalizer functions. Check back often on the HPE Developer Community blog to find more blog posts on OpsRamp."  [We're certainly hoping this will be the case!]